{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "794619ac-aa96-4785-a029-171ad69166eb",
   "metadata": {},
   "source": [
    "## Coding 1:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ee12700b-e981-44fd-8fc0-9b5eb6b54c7a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "([1, 139, 271, 502, 526, 886, 919, 1354, 1636, 1663, 2047, 2074, 2134, 2227, 2302, 2320, 2503, 2521, 2560], 19)\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "from Bio import Entrez, SeqIO\n",
    "Entrez.email = 'sahasra@uchicago.edu'\n",
    "\n",
    "class Macro():\n",
    "    def __init__(self, db, id, rettype='fasta', retmode='text'):\n",
    "        record = Entrez.efetch(db = db, id = id, rettype = rettype, retmode = retmode) #uses Entrez package to properly retrieve fasta seq info from NCBI\n",
    "        recordInfo = record.readline().split(\" \") #saves accession number, macro name and species info separately for easy access later\n",
    "        with open('%s.txt' % recordInfo[1], 'w') as file: #opens text file as name of macro for better identification of appropriate file\n",
    "            file.writelines(recordInfo)\n",
    "            self.sequence = record.read()\n",
    "            file.write(self.sequence) #writes sequence into file after writing header(which was read in first)\n",
    "        self.accessionNum = recordInfo[0]\n",
    "        self.macroType = db #saves database info as macromolecule type\n",
    "        self.macroName = recordInfo[1]\n",
    "        self.macroSpecies = recordInfo[2]\n",
    "\n",
    "    def get_macro_accessNum(self):\n",
    "        return self.accessionNum \n",
    "\n",
    "    def set_macro_name(self, newName):\n",
    "        self.macroName = newName #resets macro name in case of lack of specificity/over-specificity, or inaccuracy\n",
    "    \n",
    "    def seqLength(self):\n",
    "        return len(self.sequence) \n",
    "\n",
    "    def convertToLower(self):\n",
    "        return self.sequence.lower() #gives sequence in lowercase, if user wishes for lowercase view of seq\n",
    "\n",
    "class Protein(Macro):\n",
    "    def __init__(self, id, db=\"protein\"):\n",
    "        super().__init__(db,id) #protein subclass initialization requires specification of only 1 arg\n",
    "\n",
    "    def findOccsOfAA(self, AA=\"\"):\n",
    "        count = 0\n",
    "        AA = AA.upper()\n",
    "        for c in self.sequence:\n",
    "            if c==AA:\n",
    "                count+=1 #counts all occurrences of specified amino acid in protein seq\n",
    "        return count\n",
    "\n",
    "class Nucleotide(Macro):\n",
    "    def __init__(self, id, db='nucleotide'):\n",
    "        super().__init__(db, id) #nucleotide subclass initialization requires specification of only 1 arg\n",
    "\n",
    "    def codonCount(self):\n",
    "        cc = 0\n",
    "        for i in range(0,len(self.sequence),3):\n",
    "            cc+=1\n",
    "        return cc #returns total number of codons in nuc sequence \n",
    "    \n",
    "    def searchCodonAndOccs(self, cod):\n",
    "        codList = []\n",
    "        cod = cod.upper()\n",
    "        for i in range(0,len(self.sequence),3): #iterates frames of 3 bases\n",
    "            if self.sequence[i:i+3] == cod: #checks if any frame matches specified codon to be searched for\n",
    "                codList.append(i+1) #adds all pos(NOT index) of start of codon to list\n",
    "        return codList,len(codList) #returns list of positions as well as total number of occurrences of codon\n",
    "\n",
    "### example of use of class and subclasses with initialization of subclasses and implementation of class methods ###\n",
    "#myMacro = Macro(db='protein, id='KAI4025714.1')\n",
    "myProt = Protein(id='KAI4025714.1')\n",
    "myNuc = Nucleotide(id='AB046569.1')\n",
    "print(myNuc.searchCodonAndOccs(\"ttt\"))\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1fd67df",
   "metadata": {},
   "source": [
    "#### Macro Class Description:\n",
    "   The Macro Class defined above is to be used to fetch FASTA information about proteins, nucleotides, and more, from the NCBI database. Depending on what the user already knows about the sequence they want, they can choose to initialize an object under a specific subclass or more generically under the Macro class. The Macro class allows for easy access of information regarding the macromolecule type, the species it is from, the accession number, and the full sequence as well. In addition, the class methods gives the user the ability to quickly retrieve this information and use it to get other information, such as the length of the sequence. Depending on what the user finds more convenient, they can also choose to view the sequence in lowercase using convertToLower(). The subclasses Protein and Nucleotide show two ways in which the Macro class can be extended and defines methods that counts amino acid occurrences in a protein sequence and returns the count and positions of a codon being searched for, respectively. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "626c0fd5-993c-4269-9cb7-ef116477263b",
   "metadata": {},
   "source": [
    "## Coding 2:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5f2e5134-7c66-4380-ba62-d86468c95cd1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter a list of whole numbers separated by spaces: 2 4\n",
      "with for loop:  [4]\n",
      "with lambda:  [4]\n",
      "12.5\n",
      "12.5\n"
     ]
    }
   ],
   "source": [
    "from functools import reduce \n",
    "import random\n",
    "import math\n",
    "userIn = input(\"Please enter a list of whole numbers separated by spaces: \").split(\" \") #takes in user input of integers\n",
    "messyList = [int(i) for i in userIn] #casts each item in user input list as integer\n",
    "\n",
    "#(1)-> Using for loop:\n",
    "cleanList1 = [num for num in messyList if round(math.sqrt(num))**2==num] #uses comprehension to retain integers that are perfect squares\n",
    "\n",
    "#(1)-> Using lambda:\n",
    "cleanList2 = list(filter(lambda nums: (round(math.sqrt(nums))**2==nums), messyList)) #uses lambda to keep integers that are perfect squares\n",
    "\n",
    "print(\"with for loop: \", cleanList1)\n",
    "print(\"with lambda: \", cleanList2)\n",
    "######################\n",
    "\n",
    "randList = [random.randint(0,25) for i in range(12)]\n",
    "#(2)-> Using for loop:\n",
    "lSum = 0\n",
    "for i in randList: #uses for loop to get sum of elements in randList\n",
    "    lSum+=i\n",
    "listAvg1 = lSum/len(randList)\n",
    "print(listAvg1)\n",
    "\n",
    "#(2)-> Using lambda:\n",
    "# source = https://www.geeksforgeeks.org/python-lambda-anonymous-functions-filter-map-reduce/\n",
    "listAvg2 = (reduce(lambda i, j: i + j, randList)/len(randList)) #uses lambda and reduce to get list average by adding list elements first then dividing by list length\n",
    "print(listAvg2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3bcc3b96-7a83-4b22-acaa-648dc99836ac",
   "metadata": {},
   "source": [
    "## Coding 3:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ec226005-31fb-4b65-83a6-115fd1387cc9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Code 1a[inefficient]: \n",
      "1.616462416999994\n",
      "Code 1b[efficient]: \n",
      "2.6212622080000045\n",
      "Code 2a[inefficient]: \n",
      "3.413861374999996\n",
      "Code 2b[efficient]: \n",
      "1.704474042000001\n"
     ]
    }
   ],
   "source": [
    "import timeit #uses timeit module to note the efficiency of each version of code and check whether the changes made improved code efficiency \n",
    "# source for help with timeit: https://www.geeksforgeeks.org/timeit-python-examples/\n",
    "\n",
    "#(1)--> dot product square 2x2 matrices\n",
    "##(a: inefficient):\n",
    "mysetup = \"import random\"\n",
    "testcode = '''\n",
    "# initializes two random 2x2 matrices of integers\n",
    "myRandMat1 = [[random.randint(0,10) for i in range(2)] for y in range(2)]\n",
    "myRandMat2 = [[random.randint(0,10) for i in range(2)] for y in range(2)]\n",
    "\n",
    "#method to transpose second matrix used in dot product\n",
    "def matTranspose(mat): #source(not used exactly): https://www.geeksforgeeks.org/transpose-matrix-single-line-python/\n",
    "    tMat = []\n",
    "    for i in range(len(mat)):\n",
    "        tMat.append([])\n",
    "        for j in range(len(mat[i])):\n",
    "            currVal = mat[j][i]\n",
    "            tMat[i].append(currVal)\n",
    "    return tMat\n",
    "\n",
    "tranMat2 = matTranspose(myRandMat2)\n",
    "\n",
    "dotMat = []\n",
    "for x in range(len(myRandMat1)): #creation of dotMat containing lists for each index of final dot product matrix\n",
    "    dotMat.append([])\n",
    "    for y in range(len(tranMat2)):\n",
    "        dotMat[x].append([]) #each nested list contains products of ints that will later be summed for the final matrix\n",
    "        for z in range(len(tranMat2)):\n",
    "            dotMat[x][y].append((myRandMat1[x][z]*tranMat2[y][z])) #appends products of ints in two matrices involved in sum of each position of dot product matrix\n",
    "            #appended to nested list at correct index of dotMat list\n",
    "        \n",
    "finalMat = []\n",
    "for x in range(len(dotMat)): #iterates through nested lists and appends sum of each list to final matrix\n",
    "    finalMat.append([])\n",
    "    for y in dotMat[x]:\n",
    "        finalMat[x].append(sum(y))\n",
    "\n",
    "'''\n",
    "\n",
    "print(\"Code 1a[inefficient]: \")\n",
    "print(timeit.timeit(setup=mysetup, stmt=testcode, number=100000))\n",
    "\n",
    "##(b: efficient???):\n",
    "\n",
    "# source: https://numpy.org/doc/stable/reference/generated/numpy.dot.html\n",
    "mysetup2 = \"import numpy as np, random\"\n",
    "testcode2 = '''\n",
    "# initialization of two random 2x2 matrices of integers\n",
    "mat1 = np.random.randint(10, size=(2,2))\n",
    "mat2 = np.random.randint(10, size=(2,2))\n",
    "\n",
    "np.dot(mat1,mat2)\n",
    "'''\n",
    "print(\"Code 1b[efficient]: \")\n",
    "print(timeit.timeit(setup=mysetup2, stmt=testcode2, number=100000))\n",
    "\n",
    "\n",
    "#(2)--> sort list of lists(ascending order within each list; then ascending by first element of each list)\n",
    "##(a: inefficient):\n",
    "mysetup3 = \"import random\"\n",
    "testcode3 = '''\n",
    "\n",
    "LL = [] #random initialization of five nested lists in one main list\n",
    "for m in range(5):\n",
    "    randList = [random.randint(0,10) for i in range(random.randint(0,5))]\n",
    "    LL.append(randList)\n",
    "\n",
    "#inner sorting of lists\n",
    "for x in range(len(LL)): #iterates through inner lists\n",
    "    for y in range(len(LL[x])-1): #val1 in nested list\n",
    "        for z in range(y+1,len(LL[x])): #val2 in nested list\n",
    "            if LL[x][y] > LL[x][z]: #comparing two values\n",
    "                LL[x][y],LL[x][z] = LL[x][z],LL[x][y] #values swap positions if n > n+1\n",
    "\n",
    "for i in range(len(LL)-1):\n",
    "    for j in range(i+1,len(LL)):\n",
    "        if LL[i] > LL[j]:\n",
    "            LL[i],LL[j] = LL[j],LL[i]\n",
    "\n",
    "'''\n",
    "print(\"Code 2a[inefficient]: \")\n",
    "print(timeit.timeit(setup=mysetup3, stmt=testcode3, number=100000))\n",
    "\n",
    "##(b: efficient):\n",
    "mysetup4 = \"import random\"\n",
    "testcode4 = '''\n",
    "\n",
    "LL2 = [] #random initialization of five nested lists in one main list\n",
    "for m in range(5):\n",
    "    randList = [random.randint(0,10) for i in range(random.randint(0,5))]\n",
    "    LL2.append(randList)\n",
    "\n",
    "for l in LL2: #inner list sort\n",
    "    l.sort()\n",
    "    \n",
    "sorted(LL2) #outer list sort \n",
    "'''\n",
    "print(\"Code 2b[efficient]: \")\n",
    "print(timeit.timeit(setup=mysetup4, stmt=testcode4, number=100000))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7125d1df-9cc9-4ca6-8b97-7aa0519452bf",
   "metadata": {},
   "source": [
    "#### Coding 3(Responses)-->\n",
    "##### Explanation for efficiency/\"lack of\" efficiency in Code 1(dot product):\n",
    "   Running the inefficient and \"efficient\" code for dot product of matrices, resulted in very unexpected runtimes. While the numpy implementation of dot product was expected to run significantly quicker, it seems to run slower than the more naive implementation of matrix dot product. Where the numpy version took about 2.62 seconds to run, the naive version took 1.62 seconds to run-- a result that was very strange to say the least. After some research however, this unexpected result can be explained to be a system \"issue\". The M1 Macbook Pros were tested for data science purposes and for whatever reason (it is not known why exactly), numpy on the M1 gets executed slower when compared to the 2019 Intel system [source: https://towardsdatascience.com/are-the-new-m1-macbooks-any-good-for-data-science-lets-find-out-e61a01e8cad1]. Therefore, although this specific system runs the numpy implementation slower, it can be considered to be more efficient in general. \n",
    "\n",
    "##### Explanation for efficiency in Code 2(sort list of lists):\n",
    "   Version 2 of the nested list sorting is very obviously the more efficient method of \"two-level\" sorting. This can likely be attributed to the fact that built-in functions like .sort() and sorted() are used versus a traditional sort that involves comparing two values of a list at a time, followed by another sort using the first nested list element. The difference in time is as follows: 3.41(inefficient) and 1.70(efficient)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45185a60-b4ed-4da6-a14c-3313a7436f63",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:acs-hw1]",
   "language": "python",
   "name": "conda-env-acs-hw1-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
